### ISSUE TITLE
Pre-push hook contract is different from git's pre-push contract

### ISSUE THREAD

**Issue by samskiter** [pre-commit/pre-commit#693]
We're trying to use git-lfs with pre-commit.

Git lfs has a pre-push hook that looks like this:

```
#!/bin/sh
command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/pre-push.\n"; exit 2; }
git lfs pre-push "$@"
```

When we set this script up to be called from pre-commit's config it can hang if pre-commit uses the --all-files option (i.e. if it hits https://github.com/samskiter/pre-commit/blob/1bfd108593a268bdaf961249866f958081135ce1/pre_commit/resources/pre-push-tmpl#L13)

We can repro the issue by running pre-commit directly with --all-files. The git-lfs pre-push will hang, as if waiting for STDIN, or processing too much.

In other situations pre-commit will pass origin and source to the git-lfs pre-push script and it will run correctly.

Is there perhaps and option we could pass to pre-commit to have it run our pre-push hook as if git had called it (i.e. just fwd all arguments as if it were installed as 'legacy')?

**Comment by asottile** [comment]
hmmm I'll have to think about this one more.  pre-commit does export `PRE_COMMIT_SOURCE` and `PRE_COMMIT_ORIGIN` (which could be used to implement the inputs to `git lfs pre-push`), but as you said there are situations where pre-commit falls back to `--all-files`.

If the lfs hook is installed, it should be called [here](https://github.com/pre-commit/pre-commit/blob/51ac0e8bc0a1b61939283254c4bd5189c919f919/pre_commit/resources/hook-tmpl#L29-L32) as a legacy hook -- but of course the `stdin` / `stdout` will get exhausted by one or the other (probably needs some more clever in this situation).

I think the ideal situation would be to configure `lfs` through `pre-commit` itself using `PRE_COMMIT_ORIGIN` and `PRE_COMMIT_SOURCE`, but I'm not exactly sure what that would look like!

Thoughts?

**Comment by samskiter** [comment]
So we do have LFS configured through pre-commit with a sort of mangled version of the LFS script like so:

```
#!/bin/sh

# Check whether 'git-lfs' is installed
command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/pre-push.\n"; exit 2; }

# prePushCommand is run through Git's pre-push hook. The pre-push hook passes
# two arguments on the command line:
#
#    1. Name of the remote to which the push is being done
#    2. URL to which the push is being done
#
# Unfortunately the tool 'pre-commit' does not forward those arguments,
# thus we need to set the arguments manually in this script.
git-lfs pre-push origin git@our.git.repo/repo.git
```

This seems to completely hang when pre-commit falls back to `--all-files`

It would be great to be able to formally install a script to pre-commit as if it were legacy, without having to fiddle locally to set it up as legacy - and as you say without it consuming all of STDIN...

**Comment by asottile** [comment]
pre-commit does have an option for running legacy scripts (configured by [not specifying `--force` during `pre-commit install`](https://pre-commit.com/#running-in-migration-mode)).  Unfortunately it's just broken for anything that isn't a .git/hooks/pre-commit due to an oversight 😆 (and probably because I don't myself use the `pre-push` integration).

that said, should be an easy fix and I'll try and make a patch for this either tonight or tomorrow night when I have some free time :)

**Comment by samskiter** [comment]
Sorry, I wasn't clear. I mean it would be good to use the pre-commit config to run another script in the same way as it would be run if it were legacy, but without having to actually install that hook before installing pre-commit.

i.e. a mode where pre-commit is dumb and just forwards git's args straight to the pre-push script.

something like:

```
    -   id: run-git-lfs
        name: hooks/run_git_lfs.sh..................Runs GIT LFS
        entry: hooks/run_git_lfs.sh
        language: script
        files: .*
        always_run: true
        raw_mode: true
        stages: [push]
```

where raw_mode can just run my script as if git had run it. but explicitly without having to fiddle around copying in a pre-push.legacy file into .git (which wouldn't really work well for a shared git repo where we try to manage everything *just* using pre-commit)

**Comment by asottile** [comment]
So the current `legacy` will work if you do something like:

```bash
git lfs install  # or whatever there install command is, I forget?
pre-commit install --hook-type pre-push  # no `--force`
```

This will
1. `git-lfs` will write `.git/hooks/pre-push`
2. `pre-commit` will move `.git/hook/pre-push` => `.git/hooks/pre-push.legacy`
3. `pre-commit` will write `.git/hooks/pre-push`

The problem is, pre-commit only correctly invokes `pre-commit.legacy` (it doesn't yet know the correct semantics for passing along arguments).

I agree ideally there would be a way to run `git-lfs` from pre-commit itself, I'm not sure I want to support _two_ ways of running legacy hooks, but I'll put some thought into this.

I suspect if I change the current `--all-files` to have a source being the sigil "empty repository", then you could use git lfs's pre-push with the environment variables we expose... maybe...

(the main problem with trying to implement "legacy" pre-push is it's entirely different from the current execution approach.  pre-push is roughly equivalent to `./pre-push <<< "the args git sends"` instead of `./pre-push the args git sends`)

In summary there's two bits to this:
1. A bug in how pre-commit invokes `pre-push.legacy` / `commit-msg.legacy`
2. A potential feature request to add a "legacy"-like execution model.
    - A note that this already kind of exists for:
        - `pre-commit` (`pass_filenames: false` + `always_run: true`)
        - `commit-msg` (all executions of `commit-msg` are akin to the git calling convention).
    - pre-push is the only one where there isn't an equivalent

**Comment by asottile** [comment]
#694 addresses the `commit-msg` half of `1.` above -- `pre-push` is slightly more work than I have time for tonight :)

**Comment by asottile** [comment]
The buggy parts of this issue have been fixed and will be released in 1.6.0.

**Comment by asottile** [comment]
The bugfixes have been released as part of [v1.6.0](https://github.com/pre-commit/pre-commit/releases/tag/v1.6.0)!

**Comment by asottile** [comment]
I'm going to close this for now as "by design" -- pre-commit's current model is very centered around filenames.  That said, I could see this becoming a useful feature if done right.
